src/main/scala/io/picnicml/doddlemodel/preprocessing/QuantileDiscretizer.scala

```diff
@@ -0,0 +1,101 @@
+  * */
+object QuantileDiscretizer {
+
+  /** Create a quantile discretizer which splits data into discrete evenly sized buckets.
+    *
+    * @param bucketCounts The number of quantiles desired
+    * @param featureIndex feature index associated with features - this is needed so that only numerical features are
+    *                     transformed by this preprocessor; could be a subset of columns to be transformed
+    *
+    * @example Discretize a matrix into quartiles with two features: one numerical and one categorical.
+    *   {{{
+    *     import io.picnicml.doddlemodel.preprocessing.QuantileDiscretizer.ev
+    *
+    *     val featureIndex = FeatureIndex(List(NumericalFeature, CategoricalFeature))
+    *     val x = DenseMatrix(
+    *       List(-1.0, 0.0),
+    *       List(0.0, 1.0),
+    *       List(2.0, 0.0),
+    *       List(5.0, 0.0)
+    *     )
+    *     // equivalently, DenseVector(4) could be used
+    *     val bucketCounts = 4
+    *     val discretizer = QuantileDiscretizer(bucketCounts, featureIndex)
+    *     // Note: no fitting required
+    *     val xTransformed = ev.transform(discretizer, x)
+    *   }}}
+    */
+  def apply(bucketCount: Int, featureIndex: FeatureIndex): QuantileDiscretizer = {
+    val numNumeric = featureIndex.numerical.columnIndices.length
+    val bucketCountExtended: DenseVector[Double] = DenseVector.fill(numNumeric) { bucketCount.toDouble }
+    QuantileDiscretizer(bucketCountExtended, featureIndex)
+  }
+
+  implicit lazy val ev: Transformer[QuantileDiscretizer] = new Transformer[QuantileDiscretizer] {
+
+
+
+    @inline override def isFitted(model: QuantileDiscretizer): Boolean = model.quantiles.isDefined
+
+    override def fit(model: QuantileDiscretizer, x: Features): QuantileDiscretizer = {
+      val xCopy = x.copy
+      val discreteRangeArrays: Seq[Seq[(Double, Double)]] = model.featureIndex.numerical.columnIndices.zipWithIndex.map {
+        case (colIndex, bucketCountsIndex) =>
+          val colArray = xCopy(::, colIndex).toScalaVector.sorted
+          computeQuantiles(colArray, model.bucketCounts, bucketCountsIndex)
+      }
+      model.copy(quantiles = Some(discreteRangeArrays))
+    }
+
+    override protected def transformSafe(model: QuantileDiscretizer, x: Features): Features = {
+      val xCopy = x.copy
+      model.featureIndex.numerical.columnIndices.zipWithIndex.foreach {
+        case (colIndex, bucketsIndex) =>
+          val buckets = model.quantiles.getOrBreak(bucketsIndex)
+          (0 until xCopy.rows).foreach {
+            rowIndex => xCopy(rowIndex, colIndex) = buckets
+              .indexWhere({
+                case (lowerBound, upperBound) => lowerBound <= xCopy(rowIndex, colIndex) && xCopy(rowIndex, colIndex) <= upperBound
+              })
+              .toDouble
+          }
+      }
+      xCopy
+    }
+  }
+
+  private def computeQuantiles(target: Seq[Double], bucketCounts: DenseVector[Double], index: Int): Seq[(Double, Double)] = {
```

Comment:
Why not just take a single `bucketCount` here: `def computeQuantiles(target: Seq[Double], bucketCount: Double)` and call like this: `computeQuantiles(colArray, model.bucketCounts(bucketCountsIndex))`?

Also, `target: Seq[Double]` should be `target: Array[Double]` so there's no need to copy the data again in this function. `DenseVector[Double]` also has `.toArray` (which also copies the data I believe, but not 100% sure)

Reply:
Removed the copy, but I don't see why I would use only Array here. I can, but in my own code I use Seq almost universally in function definitions even if only one type of Seq will be used in the function in practice.